tests/user_extractor_test.php

```diff
@@ -0,0 +1,177 @@
+     */
+    public function test_get_user_by_field_from_user_table_field_does_not_exist() {
+        $this->resetAfterTest();
+
+        $user = $this->getDataGenerator()->create_user();
+
+        $this->expectException(dml_read_exception::class);
+        user_extractor::get_user('notexists', $user->username);
+    }
+
+    /**
+     * Test we can extract users using fields from {user} table when multiple users found.
+     */
+    public function test_get_user_by_field_from_user_table_when_multiple_users_found() {
+        $this->resetAfterTest();
+
+        // Two users with empty idnumber.
+        $user1 = $this->getDataGenerator()->create_user(['idnumber' => '1']);
+        $user2 = $this->getDataGenerator()->create_user(['idnumber' => '1']);
+
+        $this->expectException(more_than_one_user_found_exception::class);
+        user_extractor::get_user('idnumber', '1');
+    }
+
+    /**
+     * Test we can extract users using custom profile fields.
+     */
+    public function test_get_user_by_custom_profile_field() {
+        $this->resetAfterTest();
+
+        // Unique fields.
+        $field1 = $this->add_user_profile_field('field1', 'text', true);
+        $field2 = $this->add_user_profile_field('field2', 'text', true);
+
+        $user1 = $this->getDataGenerator()->create_user();
+        profile_save_data((object)['id' => $user1->id, 'profile_field_' . $field1->shortname => 'User 1 Field 1']);
+        profile_save_data((object)['id' => $user1->id, 'profile_field_' . $field2->shortname => 'User 1 Field 2']);
+
+        $user2 = $this->getDataGenerator()->create_user();
+        profile_save_data((object)['id' => $user2->id, 'profile_field_' . $field1->shortname => 'User 2 Field 1']);
+        profile_save_data((object)['id' => $user2->id, 'profile_field_' . $field2->shortname => 'User 2 Field 2']);
+
+        // Should find users.
+        $actual = user_extractor::get_user('profile_field_field1', 'User 1 Field 1');
+        $this->assertNotNull($actual);
+        $this->assertSame($user1->id, $actual->id);
+
+        $actual = user_extractor::get_user('profile_field_field2', 'User 1 Field 2');
+        $this->assertNotNull($actual);
+        $this->assertSame($user1->id, $actual->id);
+
+        $actual = user_extractor::get_user('profile_field_field1', 'User 2 Field 1');
+        $this->assertNotNull($actual);
+        $this->assertSame($user2->id, $actual->id);
+
+        $actual = user_extractor::get_user('profile_field_field2', 'User 2 Field 2');
+        $this->assertNotNull($actual);
+        $this->assertSame($user2->id, $actual->id);
+
+        // Shouldn't find users.
+        $actual = user_extractor::get_user('profile_field_field1', 'User 3 Field 1');
+        $this->assertNull($actual);
+
+        $actual = user_extractor::get_user('profile_field_field3', 'User 3 Field 1');
+        $this->assertNull($actual);
+    }
+
+    /**
+     * Test we can extract users using custom profile fields when found multiple users.
```

Comment:
Seems like a copy-and-paste issue here - `Test we can extract`. We actually can't

Reply:
Sorry I'm confused. What made you think it's a copy and paste issue?